packages/opentelemetry-resources/src/platform/node/detectors/AwsEcsDetector.ts

```diff
@@ -0,0 +1,97 @@
+/*
+ * Copyright The OpenTelemetry Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { Resource } from '../../../Resource';
+import { CONTAINER_RESOURCE } from '../../../constants';
+import { Detector } from '../../../types';
+import { ResourceDetectionConfigWithLogger } from '../../../config';
+import * as util from 'util';
+import * as fs from 'fs';
+import * as os from 'os';
+
+/**
+ * The AwsEcsDetector can be used to detect if a process is running in AWS
+ * ECS and return a {@link Resource} populated with data about the ECS
+ * plugins of AWS X-Ray. Returns an empty Resource if detection fails.
+ */
+export const cache = {
+  readFileAsync: util.promisify(fs.readFile),
+};
+
+class AwsEcsDetector implements Detector {
+  readonly CONTAINER_ID_LENGTH = 64;
+  readonly DEFAULT_CGROUP_PATH = '/proc/self/cgroup';
+  readonly ECS_METADATA_KEY_V4 = 'ECS_CONTAINER_METADATA_URI_V4';
+  readonly ECS_METADATA_KEY_V3 = 'ECS_CONTAINER_METADATA_URI';
+
+  async detect(config: ResourceDetectionConfigWithLogger): Promise<Resource> {
+    if (
+      !process.env.ECS_CONTAINER_METADATA_URI_V4 &&
+      !process.env.ECS_CONTAINER_METADATA_URI
+    ) {
+      config.logger.debug('AwsEcsDetector failed: Process is not on ECS');
+      return Resource.empty();
+    }
+
+    let hostName, containerId;
+    try {
+      hostName = os.hostname();
+    } catch (e) {
+      config.logger.warn(
+        `AwsEcsDetector failed to read host name: ${e.message}`
+      );
+    }
+    try {
+      const rawData = await cache.readFileAsync(
+        this.DEFAULT_CGROUP_PATH,
+        'utf8'
+      );
+      containerId = this._getContainerId(rawData);
+    } catch (e) {
+      config.logger.warn(
+        `AwsEcsDetector failed to read container ID: ${e.message}`
+      );
+    }
+
+    return !hostName && !containerId
+      ? Resource.empty()
+      : new Resource({
+          [CONTAINER_RESOURCE.NAME]: hostName || '',
+          [CONTAINER_RESOURCE.ID]: containerId || '',
+        });
+  }
+  /**
+   * Read container ID from cgroup file
+   * In ECS, even if we fail to find target file
+   * or target file does not contain container ID
+   * we do not throw an error but throw warning message
+   * and then return null string
+   */
+  private _getContainerId(rawData: string): string {
```

Comment:
This function assumes that any line longer than the container id length starts with the container id, and also will parse any line longer than this length but only use the last one.

Is there no more reliable way to find it, or a way that doesn't require iterating over every line in the file every time?

Reply:
The logic here is supposed to be read file line by line, once we find a line longer than container ID length, then break the loop immediately and return the container ID of this first valid line. 

I have added a test case for this functionality.

As for more reliable way, currently I dont think any, I will do more research about this.

Reply:
This loop will not break. You have to use a `for..of` loop in js if you want that behavior.

Reply:
```typescript
[1,2,3,4,5].forEach(n => {
  console.log(n);
  if (n == 2) return;
});

// output:
1
2
3
4
5
```

```typescript
for (const n of [1,2,3,4,5]) {
  console.log(n);
  if (n == 2) return;
};

// output:
1
2
```

Reply:
Good suggestion! Let me do it!

Reply:
In terms of a more reliable way to get the docker container ID, I might suggest getting it from `/proc/1/cpuset` (PID 1 is the init process inside the container). Something like this:

```typescript
import * as path from "path";
import * as fs from "fs";
import * as util from "util";

const readFileAsync = util.promisify(fs.readFile);

async function getContainerid() {
  const cpuset = await readFileAsync('/proc/1/cpuset');
  return path.basename(cpuset);
}
```